### ISSUE TITLE
Having issues installing/running on M1 Mac via Homebrew

### ISSUE THREAD

**Issue by kirtan** [iron-fish/ironfish#9]
/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/bindings/bindings.js:121
        throw e;
        ^

Error: dlopen(/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/segfault-handler/build/Release/segfault-handler.node, 1): no suitable image found.  Did find:
	/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/segfault-handler/build/Release/segfault-handler.node: mach-o, but wrong architecture
	/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/segfault-handler/build/Release/segfault-handler.node: mach-o, but wrong architecture
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1122:18)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at bindings (/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/opt/homebrew/Cellar/ironfish/934/libexec/node_modules/segfault-handler/index.js:3:37)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)

**Comment by erusev** [comment]
Same issue here, also on M1

**Comment by dguenther** [comment]
We probably need to look into how to ship Brew builds with M1 support, we don't have access to an M1 Mac yet. If you're able to try building/running it from source, that would be great! I suspect you'll run into an issue like this though:

```
 ironfish start

<--- Last few GCs --->

[15307:0x150008000]      198 ms: Scavenge 26.5 (33.9) -> 25.2 (33.9) MB, 0.3 / 0.0 ms  (average mu = 1.000, current mu = 1.000) allocation failure 
[15307:0x150008000]      209 ms: Scavenge 29.2 (34.4) -> 26.9 (35.1) MB, 0.3 / 0.0 ms  (average mu = 1.000, current mu = 1.000) allocation failure 
[15307:0x150008000]      225 ms: Scavenge 30.6 (35.6) -> 29.3 (37.1) MB, 0.5 / 0.0 ms  (average mu = 1.000, current mu = 1.000) allocation failure 


<--- JS stacktrace --->

FATAL ERROR: wasm code commit Allocation failed - process out of memory
```

For the Wasm out-of-memory issue, the best solution I've seen is upgrading to Node 15, so that might be something else we need to look at.

**Comment by erusev** [comment]
I tried to do it using Node 15, here's what I got, I hope it's helpful:

![image](https://user-images.githubusercontent.com/184170/113751800-f6c24900-9714-11eb-8b14-838fccd29548.png)

**Comment by yangsu** [comment]
I'm on M1 as well and here's what i'm getting using Node v14.16

![image](https://user-images.githubusercontent.com/676185/113791947-a25da000-96f9-11eb-8f1a-9eb897b76271.png)

**Comment by jesper-bylund** [comment]
Not sure if this is related, also Homebrew on M1

```bash
==> Installing ironfish from iron-fish/brew
Error: iron-fish/brew/ironfish dependencies not built for the x86_64 CPU architecture:
  pkg-config was built for arm64
  gdbm was built for arm64
  mpdecimal was built for arm64
  openssl@1.1 was built for arm64
  readline was built for arm64
  sqlite was built for arm64
  xz was built for arm64
  python@3.9 was built for arm64
```

Looks like `arm64` should be a synonym to `x86_64` to me.

**Comment by jesper-bylund** [comment]
Downloaded from GIT, changed React version from 14 to 15 in ironfish-cli and got the following error after running `yarn start` in ironfish-cli

```js
../ironfish/src/mining/miner.ts(111,29): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ type: "Buffer"; data: number[]; }' is not assignable to parameter of type 'string | { valueOf(): string; } | { [Symbol.toPrimitive](hint: "string"): string; }'.
      Property '[Symbol.toPrimitive]' is missing in type '{ type: "Buffer"; data: number[]; }' but required in type '{ [Symbol.toPrimitive](hint: "string"): string; }'.
../ironfish/src/mining/miner.ts(180,21): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ type: "Buffer"; data: number[]; }' is not assignable to parameter of type 'string | { valueOf(): string; } | { [Symbol.toPrimitive](hint: "string"): string; }'.
      Property '[Symbol.toPrimitive]' is missing in type '{ type: "Buffer"; data: number[]; }' but required in type '{ [Symbol.toPrimitive](hint: "string"): string; }'.
```

**Comment by dguenther** [comment]
M1 mac is on the way, so we should have a way to test/reproduce this ourselves soon! Thanks for tracking down the issues to get it running, that'll help a ton.

I'm not totally familiar with homebrew yet, but I think we'll need to put together a separate keg for `arm64` builds. Then homebrew should automatically download the correct one depending on what type of processor you have.

**Comment by QuickMythril** [comment]
>     3. Add `as any` to both line 111 and 180

trying to get this working on Raspberry Pi and seeing similar issues to M1 (both are ARM).  what file do you add `as any` to?

**Comment by dguenther** [comment]
It should be in this file: https://github.com/iron-fish/ironfish/blob/c4275d1e18a383f8c9b6c937bb66a81d947e7ea7/ironfish/src/mining/miner.ts#L111

**Comment by disemen** [comment]
`yarn start:once`
`yarn run v1.22.10`
`$ yarn build && yarn start:js`
`$ tsc -b`
`/bin/sh: tsc: command not found`
`error Command failed with exit code 127.`
`info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`
`error Command failed with exit code 127.`
`info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`

**Comment by dguenther** [comment]
When using https://github.com/iron-fish/ironfish/pull/244, I managed to get a node to start up on an M1 mac and sync from other nodes. I had to hack around the last remaining issues below:

* Upgrade to Node 16 for pre-built binaries that support Apple Silicon. Seems to work fine, all I had to do was bump the version requirement in `ironfish-cli/package.json`.
* `wasm-opt` is a Wasm optimization tool that doesn't distribute compatible binaries. We can bypass it with `wasm-opt = false` in the Cargo.toml, but this may have perf/size implications. At some point in the future we'll be using native node modules instead, so this will become irrelevant.

Still a ways to go, since we'll want to test #244 a bit to make sure it's not causing any stability issues, but looks promising!
